src/app/server/JointFabricDatastorage.h

```diff
@@ -0,0 +1,198 @@
+
+    CHIP_ERROR AddPendingNode(FabricIndex fabricId, NodeId nodeId, const CharSpan & friendlyName);
+    CHIP_ERROR UpdateNode(NodeId nodeId, const CharSpan & friendlyName);
+    CHIP_ERROR RemoveNode(NodeId nodeId);
+    CHIP_ERROR RefreshNode(NodeId nodeId);
+
+    CHIP_ERROR SetNode(NodeId nodeId, Clusters::JointFabricDatastore::DatastoreStateEnum state);
+
+    CHIP_ERROR AddGroupKeySetEntry(uint16_t groupKeySetId);
+    CHIP_ERROR AddGroupKeySetEntry(Clusters::JointFabricDatastore::Structs::DatastoreGroupKeySetStruct::Type & groupKeySet);
+    bool IsGroupKeySetEntryPresent(uint16_t groupKeySetId);
+    CHIP_ERROR RemoveGroupKeySetEntry(uint16_t groupKeySetId);
+    CHIP_ERROR UpdateGroupKeySetEntry(Clusters::JointFabricDatastore::Structs::DatastoreGroupKeySetStruct::Type & groupKeySet);
+
+    CHIP_ERROR AddAdmin(Clusters::JointFabricDatastore::Structs::DatastoreAdministratorInformationEntryStruct::Type & adminId);
+    bool IsAdminEntryPresent(NodeId nodeId);
+    CHIP_ERROR UpdateAdmin(NodeId nodeId, CharSpan friendlyName, ByteSpan icac);
+    CHIP_ERROR RemoveAdmin(NodeId nodeId);
+
+    const std::vector<Clusters::JointFabricDatastore::Structs::DatastoreGroupKeySetStruct::Type> & GetGroupKeySetList()
+    {
+        return mGroupKeySetList;
+    }
+    const std::vector<GenericDatastoreNodeInformationEntry> & GetNodeInformationEntries() { return mNodeInformationEntries; }
+    const std::vector<Clusters::JointFabricDatastore::Structs::DatastoreAdministratorInformationEntryStruct::Type> &
+    GetAdminEntries()
+    {
+        return mAdminEntries;
+    }
+
+    /**
+     * Used to notify of changes in the node list and more TODO.
+     */
+    class Listener
+    {
+    public:
+        virtual ~Listener() = default;
+
+        /**
+         * Notifies of a change in the node list.
+         */
+        virtual void MarkNodeListChanged() = 0;
+
+    private:
+        Listener * mNext = nullptr;
+
+        friend class JointFabricDatastorage;
+    };
+
+    /**
+     * Add a listener to be notified of changes in the Joint Fabric Datastorage.
+     *
+     * @param [in] listener  The listener to add.
+     */
+    void AddListener(Listener & listener);
+
+    /**
+     * Remove a listener from being notified of changes in the Joint Fabric Datastorage
+     *
+     * @param [in] listener  The listener to remove.
+     */
+    void RemoveListener(Listener & listener);
+
+private:
+    static constexpr size_t kMaxNodes = 32;
+
+    FabricIndex mAdministratorFabricIndex = kUndefinedFabricIndex;
+
+    NodeId mAnchorNodeId     = kUndefinedNodeId;
+    VendorId mAnchorVendorId = VendorId::NotSpecified;
+
+    std::vector<GenericDatastoreNodeInformationEntry> mNodeInformationEntries;
+    std::vector<Clusters::JointFabricDatastore::Structs::DatastoreGroupKeySetStruct::Type> mGroupKeySetList;
+    std::vector<Clusters::JointFabricDatastore::Structs::DatastoreAdministratorInformationEntryStruct::Type> mAdminEntries;
```

Comment:
I see that 3 attributes out of 13 of the JF DS Clusters are handled in this PR. Should a separate issue be opened for the rest of the work?

Reply:
Yes, this PR is an initial version and only addresses a subset of the functionalities. We’ll need follow-up PRs to complete the remaining ones.